boa/src/builtins/value/mod.rs

```diff
@@ -34,11 +35,123 @@ pub use operations::*;
 pub type ResultValue = Result<Value, Value>;
 
 /// A Garbage-collected Javascript value as represented in the interpreter.
-pub type Value = Gc<ValueData>;
+#[derive(Debug, Clone, Trace, Finalize, Default)]
+pub struct Value(pub(crate) Gc<ValueData>);
 
-#[inline]
-pub fn undefined() -> Value {
-    Gc::new(ValueData::Undefined)
+impl Value {
+    /// Creates a new `undefined` value.
+    #[inline]
+    pub fn undefined() -> Self {
+        Self(Gc::new(ValueData::Undefined))
+    }
+
+    /// Creates a new `null` value.
+    #[inline]
+    pub fn null() -> Self {
+        Self(Gc::new(ValueData::Null))
+    }
+
+    /// Creates a new string value.
+    #[inline]
+    pub fn string<S>(value: S) -> Self
+    where
+        S: Into<String>,
+    {
+        Self(Gc::new(ValueData::String(value.into())))
+    }
+
+    /// Creates a new number value.
+    #[inline]
+    pub fn rational<N>(value: N) -> Self
+    where
+        N: Into<f64>,
+    {
+        Self(Gc::new(ValueData::Rational(value.into())))
+    }
+
+    /// Creates a new number value.
+    #[inline]
+    pub fn integer<I>(value: I) -> Self
+    where
+        I: Into<i32>,
+    {
+        Self(Gc::new(ValueData::Integer(value.into())))
+    }
+
+    /// Creates a new number value.
+    #[inline]
+    pub fn number<N>(value: N) -> Self
+    where
+        N: Into<f64>,
+    {
+        Self::rational(value.into())
+    }
+
+    /// Creates a new boolean value.
+    #[inline]
+    pub fn boolean(value: bool) -> Self {
+        Self(Gc::new(ValueData::Boolean(value)))
+    }
+
+    /// Creates a new object value.
+    #[inline]
+    pub fn object(object: Object) -> Self {
+        Self(Gc::new(ValueData::Object(Box::new(GcCell::new(object)))))
+    }
+
+    /// Gets the underlying `ValueData` structure.
+    #[inline]
+    pub fn data(&self) -> &ValueData {
+        &*self.0
+    }
+
+    /// Creates a new value from a `JSONValue`.
+    #[inline]
+    pub fn from_json(json: JSONValue) -> Self {
+        Self(Gc::new(ValueData::from_json(json)))
+    }
```

Comment:
Shouldn't this be a `From<JSONValue>` iplementation?
Maybe `ValueData` should also implement `From<JSONValue>`.

Reply:
Value implements From<JSONValue>. Should ValueData implemented too?

Reply:
Hmmm if it implements `From<JSONValue>`, why do we need this function? `Value::from(json)` and `Value::from_josn(json)` would be duplicates, right?

I think it makes sense implementing it in `ValueData` too, that way `From<JSONValue>` could be like this:

```rust
fn from(json: JSONValue) -> Self {
	Self(Gc::new(json.into()))
}
```

Reply:
`Value::from` uses `Value::from_json` to convert the value, you are right it's duplicate function with the same job.
I think we should remove `Value::from_json` and only have one function.

Reply:
I think so too :)